tools/windows/install-help/cal/delfiles.cpp

```diff
@@ -122,3 +136,146 @@ BOOL DeleteFilesInDirectory(const wchar_t* dirname, const wchar_t* ext, bool dir
     
     return TRUE;
 }
+
+
+/**
+ * This function recursively deletes all files in a given tree that match a given
+ * extension.  It will only accept an absolute path.
+ *
+ * @param dirname  The root path to start the deletion
+ *
+ * @param ext   the filename and/or extension to delete.  Can be a fixed name or wildcard
+ *
+ * @param dirs  If true, will delete directories which match the ext parameter, if the
+ *              directory is empty. If false (the default), will delete files
+ */
+BOOL DeleteHomeDirectory(std::wstring &user, PSID userSID)
+{
+    // first, find the path to the home directories
+    bool ret = false;
+    wchar_t * homeDir = NULL;
+    DWORD homeDirSize = _MAX_PATH;
+    bool needsLarger = false;
+    HANDLE hFind = INVALID_HANDLE_VALUE;
+    WIN32_FIND_DATA findFileData;
+    DWORD err;
+    std::wstring search;
+    do
+    {
+        if (homeDir) {
+            delete[] homeDir;
+        }
+        homeDir = new wchar_t[homeDirSize];
+        if (!GetProfilesDirectory(homeDir, &homeDirSize))
+        {
+            err = GetLastError();
+            if (ERROR_INSUFFICIENT_BUFFER == err)
+            {
+                // loop back again.
+                needsLarger = true;
+                WcaLog(LOGMSG_STANDARD, "Finding home directory, need larger path %d", homeDirSize);
+            }
+            else {
+                WcaLog(LOGMSG_STANDARD, "Error getting home directory %d", err);
+                goto doneDelete;
+            }
+        } 
+    } while (needsLarger);
+
+    // enumerate all of the directories in the profile directory that might match
+    // this one.  Need the wildcards because the OS will add suffixes if we have a
+    // collision.
+    search = homeDir; 
+    search += L"\\*" + user + L"*";
+
+    hFind = FindFirstFile(search.c_str(), &findFileData);
+    err = GetLastError();
+    if (INVALID_HANDLE_VALUE == hFind && err != ERROR_FILE_NOT_FOUND) {
+        ret = false;
+        goto doneDelete;
+    }
+    else if (hFind != INVALID_HANDLE_VALUE)
+    {
+        for (bool bContinue = true; bContinue; bContinue = FindNextFile(hFind, &findFileData))
```

Comment:
Your method won't work.  using the while method as you suggest will skip the first entry, because `FindNextFile()` is called on entry into the loop.  The `for` method doesn't call FNF until the end of the loop, so the first iteration executes on the return from `FindFirstFile()` above, as expected.

The alternative if you prefer `while` is to do `do.... while()` with your contitional at the end.